### ISSUE TITLE
Spurious Fiber Failure Warning When Fiber#orElse is Used

### ISSUE THREAD

**Issue by adamgfraser** [zio/zio#2349]
From [this](https://stackoverflow.com/questions/59063170/zio-fiber-orelse-generate-exception-messages) Stack Overflow question. The code below generates a fiber failure warning beause `f1f` fails but it is spurious because it is anticipated and handled by `Fiber#orElse`.

```scala
import zio._
import zio.console._

object MyApp extends App {

  def f1 :Task[Int] = IO.fail(new Exception("f1 fail"))
  def f2 :Task[Int] = IO.succeed(2)

  val myAppLogic =
    for {
      f1f <- f1.fork
      f2f <- f2.fork
      ff  =  f1f.orElse(f2f)
      r   <- ff.join
      _   <- putStrLn(s"Result is [$r]")
    } yield ()

  def run(args: List[String]) =
    myAppLogic.fold(_ => 1, _ => 0)
}
```

**Comment by adamgfraser** [comment]
@Twizty Do you want to fix as part of #2309 or should we handle this separately?

**Comment by jdegoes** [comment]
@adamgfraser Hmm, this is an important use case, generated by `Fiber#orElse`.

The other ones seem solvable but this one seems trickier since the combinator only has the output of another fiber to go on. Hmm...

**Comment by adamgfraser** [comment]
@jdegoes Yes, I think we are seeing that not only do we need the ability to mark a fiber as as not reporting errors at creation but also have the ability to add that status later. I think that might point towards your proposal of fibers containing handlers because then you could provide a method on `Fiber` that changes the handler in `FiberContext`, whereas I'm not sure how we handle that in the "error determines how it should be reported" model because all the combinators on just change the value when it is awaited or joined which never happens here.

**Comment by Twizty** [comment]
@jdegoes @adamgfraser I don't think it is possible to catch an error of a fiber that immediately fails, and I don't know if it is a good idea to mark fibers as "evaluated later". How to determine when that "later" will be?

**Comment by jdegoes** [comment]
@Twizty @adamgfraser 

What we are really looking for is the following:

 - Do not log the unhandled error if someone else is going to see it

And indeed, this is the problem: one can see the error only after calling `join` / `await` / `poll`, by which point, the fiber may have already finished.

So fibers don't have enough information when they end life to determine if someone saw the error.

Even if we add hierarchical supervision (which I think is maybe a good idea), we will still be faced with this fundamental problem. 

I see three possible solutions:

 - **Push the solution onto the users**. When you fork, you get to specify whether or not to log unhandled errors (this is equivalent with saying, when you get to fork, you can specify a custom unhandled error reporter, which can piggyback atop hierarchical supervision). For our own ZIO combinators, this will work fine, because we know when we fork if we will ever join, thus eliminating the potential race condition. This solution does not work with `Fiber#orElse`.
 - **By default, only log defects**. When someone forks an effect that may fail, they have static information that the effect they are forking may fail, and they have chosen not to deal with that failure. So one can argue that typed errors, together with interruptions, should not be logged because the user already could see them and could have chosen to do something about them. For convenience, we could add something like `forkWithErrorHandler((e: E) => log(e) : UIO[Unit])` which simply calls `.run` on the effect and passes errors to the error handler, before forking it, and then continuing with `IO.done`.  This solution works with `Fiber#orElse` for typed errors and interruptions.
 - **Buffer unhandled errors before logging them**. By choosing some window of time before logging unhandled errors, e.g. 1 second, one can heuristically detect whether or not an error was handled. An error was not handled if, within some bounded time window, no one saw it. This has the nice property of eliminating spurious errors, but it cannot eliminate all spurious errors, since no matter what time window you choose, there will be rare circumstances when beyond that window, a fiber is joined, resulting in a spurious log message. This solution works with `Fiber#orElse` for all errors. The question with this solution is: just because someone saw the unhandled error, does that necessarily imply we don't want to log it? 

For the log to be useful, ultimately we should only put things there if the user should really take a look at it, so it makes sense to be much more conservative than we are being now. Users are confused by all these log messages that don't really require any attention and don't occur with other effect systems.

**Comment by ghostdogpr** [comment]
My preference goes to **By default, only log defects**. Those are the only ones that shouldn’t happen and deserve to be logged by default. Also it keeps things simple.

**Comment by adamgfraser** [comment]
Closed by #2339.
